ts/main.ts

```diff
@@ -194,56 +239,175 @@ async function renderAllCardsInitial(): Promise<void> {
   cardsArray.forEach((element) => {
     $spellsListCardsDiv.appendChild(element);
   });
 }
 
+// event listener for when the sort order is changed
 $spellsListSortDropdown.addEventListener('input', () => {
-  // clear cards
-  while ($spellsListCardsDiv.childNodes.length > 0) {
-    if (!$spellsListCardsDiv.firstElementChild) break;
-    $spellsListCardsDiv.removeChild($spellsListCardsDiv.firstElementChild);
-  }
+  // get the sort value from the dropdown
+  cardSort.sort = $spellsListSortDropdown.value;
 
   // re-render cards based on sort value
-  if ($spellsListSortDropdown.value === 'level') {
-    sortCardsLevel();
+  sortCards(cardSort.sort);
+});
+
+// event listener for when the search bar is submitted
+$spellsListSearchSortForm.addEventListener('submit', (event: Event) => {
+  event.preventDefault();
+
+  cardSort.filter.name = $spellsListSearchInput.value;
+  $spellsListFilterNameInput.value = cardSort.filter.name;
+
+  if ($spellsListSearchInput.value) {
+    filterSpellsList();
   } else {
-    sortCardsName();
+    for (let i = 0; i < cardsArray.length; i++) {
+      cardsArray[i].classList.remove('hidden');
+    }
+  }
+});
+
+// event listener for when the 'filter' button is clicked
+$spellsListFilterBtn.addEventListener('click', () => {
+  $spellsListFilterDialog.showModal();
+});
+
+/* function to query the api for a filter endpoint, and show or hide cards in
+the card array depending on if its name is returned from the fetch */
+async function filterSpellsList(): Promise<void> {
+  try {
+    // creates an endpoint url for the fetch command
+    // based on the values in the cardSort object
+    let apiFilterUrl = '?';
+    if (cardSort.filter.name) {
+      const urlName = cardSort.filter.name.replace(' ', '%20');
+      apiFilterUrl += `name=${urlName}`;
+    }
+    if (cardSort.filter.level > -1) {
+      apiFilterUrl += `&level=${cardSort.filter.level}`;
+    }
+    if (cardSort.filter.school) {
+      apiFilterUrl += `&school=${cardSort.filter.school}`;
+    }
+    const response = await fetch(
+      `https://www.dnd5eapi.co/api/spells${apiFilterUrl}`,
+    );
+
+    if (!response.ok) throw new Error(`Fetch error status: ${response.status}`);
+
+    // gets the API response as an object, then loops through to append all
+    // the spell names to a new array
+    const filteredSpellData = (await response.json()) as AllSpells;
+    const filteredSpellNames: string[] = [];
+    filteredSpellData.results.forEach((element) => {
+      filteredSpellNames.push(element.name);
+    });
+
+    // loops through the card array and shows or hides depending if its name
+    // is included in the array of names
+    cardsArray.forEach((element: HTMLDivElement) => {
+      const elementName = element.getAttribute('data-name') as string;
+      if (!elementName) return;
+      if (filteredSpellNames.includes(elementName)) {
+        element.classList.remove('hidden');
+      } else {
+        element.className += ' hidden';
+      }
```

Comment:
Minor thing but if you used `classList` on 269 I would use `classList` on 271 for consistency.
`element.classList.add('hidden')`